### PR TITLE
Getting Started Guide

### PR DESCRIPTION
**Pull request by brooksmtownsend**
Initial draft of the "Getting Started" guide, including a fully guided example of starting the `Echo` actor and `httpserver` provider to show a basic development flow.

@autodidaddict Looking for any feedback on tone, style of writing, etc here. I haven't written much technical documentation, and I feel my tutorials err on the casual side so let me know if this fits with the rest of the docs.

@billyoung I might need you to fill me in on the Homebrew steps for mac, happy to test them out.

As a part of this PR I beefed up the installation guide to include a Macos section, and to add an "install from source" section.

The only thing we need to do to make this guide fully work is to publish the echo actor to our azurecr. This can be done manually but I'm fine doing it VIA a GH action in the `examples` repo once that's available since that's a prereq to "big release"

### CODE REVIEW

File: content/overview/getting-started/_index.en.md [hunk 1]
```diff
@@ -5,50 +5,149 @@ weight: 2
 draft: false
 ---
 
-### Global site parameters
+In this guide, we'll be covering the following starting actions for wasmCloud development:
```
Comment (autodidaddict) [inline content/overview/getting-started/_index.en.md:8]: The phrase "getting started" here I thought was a bit more figurative. Rather than telling people "how to start all the things", this is really just the first steps you need to take as a developer to get started on your journey of learning how to develop wasmCloud applications. 

For example, starting an actor is covered in the creating an actor guide. Starting a provider is covered in the creating a provider guide, etc.

Comment (autodidaddict) [reply content/overview/getting-started/_index.en.md:8]: Maybe instead of "the following starting actions", we could say that in this guide, we're going to be taking a **tour** of some of the most common activities in the wasmCloud ecosystem, like starting and configuring actors and providers.

I think the "tour" phrase here gives people a better perspective that you're going to walk them through things in this guide that are also discussed at length elsewhere in the docs.

File: content/overview/getting-started/_index.en.md [hunk 2]
```diff
@@ -5,50 +5,149 @@ weight: 2
...
 
-On top of [Hugo global configuration](https://gohugo.io/overview/configuration/), **Dot** lets you define the following parameters in your `config.toml` (here, values are default).
+### Prerequisites
+In order to follow this guide, you'll need a few things:
+- `wash` (installation covered on [the previous page](../installation))
+- [nats-server](https://docs.nats.io/nats-server/installation) **OR** [docker](https://docs.docker.com/engine/install/), recommended with [docker-compose](https://docs.docker.com/compose/install/)
 
-Note that some of these parameters are explained in details in other sections of this documentation.
+### Starting NATS
+[NATS](https://nats.io/) is a reliable cross-environment messaging system that wasmCloud uses extensively to send messages between the core program components. wasmCloud _can_ be run without using NATS, however this is only recommended in extremely constrained environments, or when NATS is unavailable.
```
Comment (autodidaddict) [inline content/overview/getting-started/_index.en.md:21]: This might be a bit overwhelming at this stage of the developer journey (since we're assuming they haven't read any reference guides yet). Maybe "NATS is a message broker used by wasmCloud's self-managing **lattice** network" and we leave  it at that (and possibly make the word lattice a link to the lattice reference).

File: content/overview/getting-started/_index.en.md [hunk 3]
```diff
@@ -5,50 +5,149 @@ weight: 2
...
+   {{% tab "nats-server binary" %}}
+    nats-server -a 0.0.0.0 -p 4222
+   {{% /tab %}}
+ {{% /tabs %}}
 
-# add css plugin
-[[params.plugins.css]]
-link = "define plugins path"
+### Running your first wasmCloud Host
+In a separate terminal window, run the following command to launch an interactive wasmCloud REPL environment, including a preconfigured wasmCloud host.
```
Comment (autodidaddict) [inline content/overview/getting-started/_index.en.md:57]: typical convention leading up to a code block and we use the word/phrase "the following" is to end this sentence with a `:` instead of a `.`. 

I'm not sure which AP style guide that comes from, but it's a thing ;)

File: content/overview/getting-started/_index.en.md [hunk 4]
```diff
@@ -5,50 +5,149 @@ weight: 2
...
+```shell
+wash up
+```
+At any time, you can type `help` to see a printed list of subcommands in the log output window in the bottom-right pane. To exit the repl, you can type `exit`, `logout`, `q` or `:q!`.
+![REPL](../wash_up.png)
```
Comment (autodidaddict) [inline content/overview/getting-started/_index.en.md:62]: Do we need to give some recommendations for how much room they need in a terminal prompt? I'm assuming they need to be able to provide at least 80x25 characters? When I did this in a small window, I couldn't see the help text.

File: content/overview/getting-started/_index.en.md [hunk 5]
```diff
@@ -5,50 +5,149 @@ weight: 2
...
+hostcore.arch                                                    x86_64                                       
+repl_mode                                                        true                                         
+                          No actors found                                             
+                                                                                                        
+Provider ID               Link Name                  Image Reference   
+VDHPKGFKDI34Y4RN4PWWZH... default                    N/A               
+VAHNM37G4ARHZ3CYHB3L34... __wasmcloud_lattice_cache  N/A               
+```
+Currently on this Host, we have a few labels that show the environment this Host is running on, and two default providers. The first, starting with `VDHPK`, is the `extras` provider, which provides an interface for wasmCloud resources to generate random numbers, GUIDS, etc. The second, starting with `VAHNM`, is the lattice cache provider, which provides a distributed cache to all hosts running in this lattice. We'll talk more about what a **Lattice** is in the [Platform Builder](../../platform-builder) section of the documentation.
```
Comment (autodidaddict) [inline content/overview/getting-started/_index.en.md:86]: `Host` here shouldn't be capitalized.

Comment (autodidaddict) [reply content/overview/getting-started/_index.en.md:86]: We can link `lattice` here to the reference guide documentation on lattice.

File: content/overview/getting-started/_index.en.md [hunk 6]
```diff
@@ -5,50 +5,149 @@ weight: 2
...
 
+#### Running an Actor
+We can start scheduling Actors and Providers right away on this Host using the `ctl start` command. To begin, we'll start our `Echo` sample actor.
```
Comment (autodidaddict) [inline content/overview/getting-started/_index.en.md:89]: I've gone back and forth on this, and editorial style and copy editing say that actors and providers shouldn't be capitalized.

Comment (autodidaddict) [reply content/overview/getting-started/_index.en.md:89]: Host here is capitalized and doesn't need to be.

File: content/overview/getting-started/_index.en.md [hunk 7]
```diff
@@ -5,50 +5,149 @@ weight: 2
...
+```
+The `Echo` sample actor has a single operation `HandleRequest`, which will respond to a delivered HTTP Request with an identical response that "echoes" the request sent.
 
-# main menu
-[[menu.main]]
-name = "contact"
-url = "contact"
-weight = 1
+#### Running a Capability Provider
+In order for this actor to receive HTTP requests, we're going to need to start the HTTP Server Capability Provider. Actors are signed WebAssembly modules that do not have access to create or receive HTTP requests, so the capability provider will do that for us and deliver messages to and from the actor.
```
Comment (autodidaddict) [inline content/overview/getting-started/_index.en.md:96]: this is confusing to me, the `do not have access` bit makes me think actors don't have access to HTTP providers, but they do once we sign them. Maybe rephrase this a bit so it's a bit more clear that "Actors are signed WebAssembly modules, and they can have embedded claims declaring their ability to communicate with capability providers like the HTTP server" (or whatever)

File: content/overview/getting-started/_index.en.md [hunk 8]
```diff
@@ -5,50 +5,149 @@ weight: 2
...
-[[menu.main]]
-name = "contact"
-url = "contact"
-weight = 1
+#### Running a Capability Provider
+In order for this actor to receive HTTP requests, we're going to need to start the HTTP Server Capability Provider. Actors are signed WebAssembly modules that do not have access to create or receive HTTP requests, so the capability provider will do that for us and deliver messages to and from the actor. 
+```shell
+wash> ctl start provider wasmcloud.azurecr.io/httpserver:0.10.0
+```
+Let's take a look at our host's inventory now, you can either type the command above again or use the `UP/DOWN` arrow keys to navigate to the previous command. It should look something like this (some output truncated for documentation):
```
Comment (autodidaddict) [inline content/overview/getting-started/_index.en.md:100]: "inventory now. You can either"

File: content/overview/getting-started/_index.en.md [hunk 9]
```diff
@@ -5,50 +5,149 @@ weight: 2
...
+Provider ID    Link Name        Image Reference                        
+VDHPKGFKDI3... default          N/A                                    
+VAHNM37G4AR... __wasmcloud_...  N/A                                    
+VAG3QITQQ2O... default          wasmcloud.azurecr.io/httpserver:0.10.0 
+```
 
-# Call to action is default enabled, if you want to disable it. just change the 
-enable = false
+#### Linking Actors and Capability Providers
+Now your `Echo` Actor and `HTTP Server` Providers are running, but they aren't connected. In order to allow the Actor and Provider to communicate, we need to `link` them together, which we do using their respective Public Keys (shown in the Inventory as `Actor ID` and `Provider ID`).
```
Comment (autodidaddict) [inline content/overview/getting-started/_index.en.md:119]: actor
providers
public keys
all lower case.

File: content/overview/getting-started/_index.en.md [hunk 10]
```diff
@@ -5,50 +5,149 @@ weight: 2
...
-[params]
-logo = "images/logo.png"
-# Meta data
-description = "This is meta description"
-author = "Themefisher"
-# contact form action
-contact_form_action = "#" # contact form works with https://formspree.io
-# Google Analitycs
-google_analitycs_id = "" # your id
+You can use the following command to `link` your Actor and Provider, on MacOS you might receive a message asking if you want `wash` to receive incoming network connections. You should click "Allow", this is just setting up the provider's local HTTP server to listen on port 8080.
```
Comment (autodidaddict) [inline content/overview/getting-started/_index.en.md:121]: your actor and provider. On macOS you might...

Comment (autodidaddict) [reply content/overview/getting-started/_index.en.md:121]: Might want to be explicit that the dialog box on the mac is creating a temporary firewall exception.

File: content/overview/getting-started/_index.en.md [hunk 11]
```diff
@@ -5,50 +5,149 @@ weight: 2
...
+{"method":"GET","path":"/echo","query_string":"","headers":{"host":"localhost:8080","user-agent":"curl/7.64.1","accept":"*/*"},"body":[]}
+```
+Feel free to try out different methods of making a request to your Actor, including adding headers or using a different HTTP method to see different outputs.
+
+Instead of using `curl`, you can also _directly invoke_ actors' registered functions using `ctl call`. As mentioned before, the function that "echoes" this HTTP request is registered with the operation `HandleRequest`, and we can make this request directly to the actor if we supply the correct parameters.
+
+Here's an example of using `call` in the REPL to mimic our above `curl`. Note that this is not interacting with the HTTP server provider, and we don't need it to be running or linked for this operation to succeed:
+```shell
+wash> ctl call MBCFOPM6JW2APJLXJD3Z5O4CN7CPYJ2B4FTKLJUR5YR5MITIU7HD3WD5 HandleRequest {"method": "GET", "path": "/echo", "body": "", "queryString":"","header":{}}
+```
```
Comment (autodidaddict) [inline content/overview/getting-started/_index.en.md:143]: Might want to use a caution/warning emoji here and drop a call-out explaining that this only works because the JSON payload you're passing is _exactly_ the same in terms of fields, shape, and data types as the messagepack payload the actor is expecting. If they don't match this up correctly, the actor will reject the call.

File: content/overview/getting-started/_index.en.md [hunk 12]
```diff
@@ -5,50 +5,149 @@ weight: 2
...
+Our output will look something like this:
+```shell
+Call response (raw): ��statusCode�Ȧstatus�OK�header��body�S{"method":"GET","path":"/echo","query_string":"","headers":{},"body":[]}
+```
+Because of our method of serialization / deserialization, the return payload has some characters that the terminal doesn't know how to interpret from its bytes.[<sup>1</sup>](#references) However you can still see the response body which contains our exact "echoed" request.
```
Comment (autodidaddict) [inline content/overview/getting-started/_index.en.md:148]: maybe change "Because of our method of serialization  / deserialization" we can just say "Because the actor isn't actually returning JSON"...

File: content/overview/getting-started/_index.en.md [hunk 13]
```diff
@@ -5,50 +5,149 @@ weight: 2
...
+
+That covers the basics of wasmCloud development in our REPL environment. To learn more about Actors, Providers, and more concepts on wasmCloud, continue onto the [App Development](../../app-dev) or [Platform Building](../../platform-builder) sections depending on your role.
```
Comment (autodidaddict) [inline content/overview/getting-started/_index.en.md:150]: s/onto/on to

maybe "depending on your role" could change to "depending on your interests" ("role" feels like a limiting phrase here and we want to be welcoming in intro guides)

File: content/overview/getting-started/_index.en.md [hunk 14]
```diff
@@ -5,50 +5,149 @@ weight: 2
...
+
+###### References
+1. We're using [msgpack](https://msgpack.org/) to serialize arbitrary byte payloads when sending information to/from actors. Due to this serialization format, the bytes that are sent to and from actors must be deserialized with the shape of the object already known, and in the case of `ctl call` we are doing our best guess as to what the return payload is going to look like. As you can see when using `curl`, when we know the shape of the data (an HTTP response) we're able to fully deserialize the actor's response.
```
Comment (autodidaddict) [inline content/overview/getting-started/_index.en.md:153]: If this is a footnote, we can actually use real markdown footnote syntax, the hugo template will even draw a little gray hard rule above your footnotes section.

**Review by autodidaddict — CHANGES_REQUESTED** [review]

File: content/overview/getting-started/_index.en.md (hunk 4 above)
Comment (brooksmtownsend) [reply content/overview/getting-started/_index.en.md:62]: Yeah I think we do need to give some recommendations, especially with width as the table output can appear strange. I'll play around with the `help` output height and `get inventory` width to come up with a recommended size. Perhaps in future versions of `wash` we could include an initial character output in the Output window to help them size correctly

File: content/overview/getting-started/_index.en.md (hunk 3 above)
Comment (brooksmtownsend) [reply content/overview/getting-started/_index.en.md:57]: That makes sense, I believe I did that in other places too but I'll comb through and make that consistent

File: content/overview/getting-started/_index.en.md (hunk 2 above)
Comment (brooksmtownsend) [reply content/overview/getting-started/_index.en.md:21]: Perfect, it felt like a place where I needed to say more than "just start this binary for no reason" and less than exactly what NATS is used for.

File: content/overview/getting-started/_index.en.md (hunk 1 above)
Comment (brooksmtownsend) [reply content/overview/getting-started/_index.en.md:8]: I like `tour` too, since we aren't requiring any action by the developer other than copy/pasting the host ID. everything else is stubbed out

File: content/overview/getting-started/_index.en.md (hunk 6 above)
Comment (brooksmtownsend) [reply content/overview/getting-started/_index.en.md:89]: re: this and above `Host` comment, I'll convert the Host, Actor, and Provider mentions to be lowercased.

File: content/overview/getting-started/_index.en.md (hunk 14 above)
Comment (brooksmtownsend) [reply content/overview/getting-started/_index.en.md:153]: Sweet I didn't know there was an official way to do this. On it

File: content/overview/installation/_index.en.md [hunk 15]
```diff
@@ -39,3 +39,10 @@ You can find our packagecloud.io page [here](https://packagecloud.io/wasmCloud/c
    {{% /tab %}}
 {{% /tabs %}}
 
+### (MacOS) Install the Homebrew tap
+wasmCloud uses [Homebrew](https://brew.sh/) for our MacOS builds
+
+//TODO
```
Comment (autodidaddict) [inline content/overview/installation/_index.en.md:45]: If this is no longer a todo, we should remove it from the doc.

**Review by autodidaddict — CHANGES_REQUESTED** [review]

File: content/overview/installation/_index.en.md (hunk 15 above)
Comment (brooksmtownsend) [reply content/overview/installation/_index.en.md:45]: @billyoung Are the homebrew instructions ready? If not, I can remove it and we can add it later.

Comment (billyoung) [reply content/overview/installation/_index.en.md:45]: @brooksmtownsend the instructions are here: https://github.com/wasmCloud/wasmcloud-dev-site/blob/main/content/overview/installation/_index.en.md

```
brew tap wasmcloud/wasmcloud
brew install wasmcloud wash
```
Just a reminder: there will not be a homebrew package for `wasmCloud` until we create a release

Comment (brooksmtownsend) [reply content/overview/installation/_index.en.md:45]: Going to throw these in the install doc then, and we'll make sure to release it 😛

**Review by autodidaddict — APPROVED** [review]
